python/cuml/feature_extraction/text.py

```diff
@@ -0,0 +1,578 @@
+        tokenized_df = tokenized_df.reset_index(drop=True)
+
+        return tokenized_df
+
+    def _insert_zeros(self, ary, zero_indices):
+        """Create a new array of len(ary + zero_indices) where zero_indices
+        indicates indexes of 0s in the new array. Ary is used to fill the rest.
+
+        Example:
+            _insert_zeros([1, 2, 3], [1, 3]) => [1, 0, 2, 0, 3]
+        """
+        if len(zero_indices) == 0:
+            return ary.values
+
+        new_ary = cp.zeros((len(ary) + len(zero_indices)), dtype=cp.int32)
+
+        # getting mask of non-zeros
+        data_mask = ~cp.in1d(cp.arange(0, len(new_ary)), zero_indices)
+
+        new_ary[data_mask] = ary
+        return new_ary
+
+    def _compute_empty_doc_ids(self, count_df, n_doc):
+        """
+        Compute empty docs ids using the remaining docs, given the total number
+        of documents.
+        """
+        remaining_docs = count_df['doc_id'].unique().values
+        doc_ids = cp.arange(0, n_doc)
+        empty_doc_ids = doc_ids[~cp.in1d(doc_ids, remaining_docs)]
+        return empty_doc_ids
+
+    def _create_csr_matrix_from_count_df(self, count_df, empty_doc_ids):
+        """Create a sparse matrix from the count of tokens by document"""
+        n_features = len(self.vocabulary_)
+
+        data = count_df["count"].values
+        indices = count_df["token"].values
+
+        doc_token_counts = count_df["doc_id"].value_counts().reset_index()
+        doc_token_counts = doc_token_counts.rename(
+            {"doc_id": "token_counts", "index": "doc_id"}
+        ).sort_values(by="doc_id")
+        token_counts = self._insert_zeros(doc_token_counts["token_counts"],
+                                          empty_doc_ids)
+        indptr = token_counts.cumsum()
+        indptr = cp.pad(indptr, (1, 0), "constant")
+
+        n_rows = len(doc_token_counts) + len(empty_doc_ids)
+
+        return cp.sparse.csr_matrix(
+            arg1=(data, indices, indptr), dtype=self.dtype,
+            shape=(n_rows, n_features)
+        )
+
+    def _validate_params(self):
+        """Check validity of ngram_range parameter"""
+        min_n, max_m = self.ngram_range
+        msg = ""
+        if min_n < 1:
+            msg += "lower boundary must be >= 1. "
+        if min_n > max_m:
+            msg += "lower boundary larger than the upper boundary."
+        if msg != "":
+            msg = f"Invalid value for ngram_range={self.ngram_range} {msg}"
+            raise ValueError(msg)
+
+    def _warn_for_unused_params(self):
+        if self.analyzer != 'word' and self.stop_words is not None:
+            warnings.warn("The parameter 'stop_words' will not be used"
+                          " since 'analyzer' != 'word'")
+
+
+def _document_frequency(X):
+    """Count the number of non-zero values for each feature in X."""
+    doc_freq = cp.asarray(
+        X[["token", "doc_id"]]
+        .groupby(["token"])
+        .count()
+        .as_gpu_matrix()
+    ).ravel()
```

Comment:
This might be expensive , any reason not to do below instead ?  
```python
doc_freq = X[["token", "doc_id"]].groupby(["token"]).count().doc_id.values

```

Reply:
No you are right, I didn't realize I could use `.values` instead of `.as_gpu_matrix()` here TBH.